### PR TITLE
Fixed meme

### PR DESCRIPTION
**Pull request by allopez7**

### CODE REVIEW

File: Meme/express.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+app.use(bp.json({limit: "500mb"}));
+
+const response = (req,res) =>{
+  fs.appendFile('/home/alopez/garagescript/Meme/public/book.txt',`${req.body.name}: ${req.body.comment}`+'\n',()=>{});
+  res.send();
+}
+app.post('/send', response);
+
+const resp2 = (req, res) =>{
+  fs.writeFile('/home/alopez/garagescript/Meme/public/'+req.body.name+'.'+'png',req.body.Canvas.replace('data:image/png;base64',""),'base64',() =>{
```
Comment (songz) [inline Meme/express.js]: don't add strings by using + anymore. use ```

File: Meme/express.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+   gm('/home/alopez/garagescript/Meme/public/'+req.body.name+'.'+'png').fontSize(40).drawText(50,50,req.body.comment).write('/home/alopez/garagescript/Meme/public/text.png',(err)=>{
```
Comment (songz) [inline Meme/express.js]: same here. You are using same path here, so rather than being repetitive, please use a variable.

File: Meme/express.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+app.listen(3608);
+app.use(express.static('public'));
+app.use(bp.json({limit: "500mb"}));
+
+const response = (req,res) =>{
+  const chatMessage = `${req.body.name}: ${req.body.comment}\n`;
+  const textFilePath = '/home/alopez/garagescript/Meme/public/book.txt';
+
+  fs.appendFile(textFilePath, chatMessage,()=>{});
+  res.send();
```
Comment (chocobrownie) [inline Meme/express.js]: Can you try to write something like res.send("name and comment written in /book.txt")  ?

File: Meme/express.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+const express = require('express');
+const bp = require('body-parser');
+const app = express();
+const gm = require('gm');
+
+app.listen(3608);
+app.use(express.static('public'));
+app.use(bp.json({limit: "500mb"}));
+
+const response = (req,res) =>{
```
Comment (chocobrownie) [inline Meme/express.js]: what does "response" mean as a variable name?

File: Meme/express.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  const chatMessage = `${req.body.name}: ${req.body.comment}\n`;
+  const textFilePath = '/home/alopez/garagescript/Meme/public/book.txt';
+
+  fs.appendFile(textFilePath, chatMessage,()=>{});
+  res.send();
+}
+app.post('/send', response);
+
+const resp2 = (req, res) =>{
```
Comment (chocobrownie) [inline Meme/express.js]: same for resp2

Comment (chocobrownie) [reply Meme/express.js]: Actually, this is not a response, this is a url component that handles request and gives back response. So by concept this is not correct, plus the purpose of this function is not super readable

File: Meme/express.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  const imagePath = `/home/alopez/garagescript/Meme/public/${req.body.name}.png`;
+  const imageData = req.body.Canvas.replace('data:image/png;base64',""); 
+  const imageMeme = '/home/alopez/garagescript/Meme/public/text.png'; 
+
+  fs.writeFile(imagePath, imageData,'base64',() =>{
+   gm(imagePath).fontSize(40).drawText(50,50,req.body.comment).write(imageMeme,(err)=>{
```
Comment (chocobrownie) [inline Meme/express.js]: if you convert every image into one file text.png then you end up only have one meme, which is not the purpose of this lesson. The challenge asks you to generate meme for every user and display it.

File: Meme/express.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+     console.log("error is", err);
```
Comment (chocobrownie) [inline Meme/express.js]: error should only be printed when there is an error, not in every circumstances

File: Meme/express.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+   });
+});
+  res.send('done');
```
Comment (chocobrownie) [inline Meme/express.js]: try to send a clearer message of what it did

File: Meme/public/index.html [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+} 
+
+window.setInterval(getTextFile, 1000);
+
+const video = document.getElementById('video');
+
+navigator.mediaDevices.getUserMedia({video: true, audio: false}).then((source) =>{ video.srcObject = source;
+  video.play();
+});
+
```
Comment (chocobrownie) [inline Meme/public/index.html]: where is your code to display the memes?

Comment (allopez7) [reply Meme/public/index.html]: I'm not sure what you mean here. I don't think anything is required here for meme.

File: Meme/public/index.html [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+  httpReq2.send();
+  
+} 
+
+window.setInterval(getTextFile, 1000);
+
+const video = document.getElementById('video');
+
+navigator.mediaDevices.getUserMedia({video: true, audio: false}).then((source) =>{ video.srcObject = source;
```
Comment (songz) [inline Meme/public/index.html]: Bring to new line!
```
(source) => {
  video.srcObject...
  video.play();
}
```

File: Meme/public/index.html [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+  const name = document.getElementById('name').value; 
+  const comment = document.getElementById('comment').value;
+  const context = canvas.getContext('2d');
+  const httpReq = new XMLHttpRequest();
+  context.drawImage(video,0,0);
+  const dataURL = canvas.toDataURL('image/png');
+  
+  httpReq.open('POST','/image');
+  httpReq.setRequestHeader('content-type','application/json');
+  httpReq.send(
```
Comment (songz) [inline Meme/public/index.html]: better syntax:
```
httpReq.send(JSON.stringify({
  Canvas: dataURL, name, comment
}));
```

File: Meme/express.js [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+
+const createImageMeme = (req, res) =>{
+  const imagePath = `/home/alopez/garagescript/Meme/public/${req.body.name}.png`;
+  const imageData = req.body.Canvas.replace('data:image/png;base64',""); 
+  const imageMeme = `/home/alopez/garagescript/Meme/public/${req.body.name}.png`; 
+
+  fs.writeFile(imagePath, imageData,'base64',() =>{
+   gm(imagePath).fontSize(40).drawText(50,50,req.body.comment).write(imageMeme,()=>{
+   });
+});
```
Comment (songz) [inline Meme/express.js]: indent!

File: Meme/express.js [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  const textFilePath = '/home/alopez/garagescript/Meme/public/book.txt';
+
+  fs.appendFile(textFilePath, chatMessage,()=>{});
+  res.send('append comment to /book.txt');
+}
+app.post('/send', appendTextToFile);
+
+const createImageMeme = (req, res) =>{
+  const imagePath = `/home/alopez/garagescript/Meme/public/${req.body.name}.png`;
+  const imageData = req.body.Canvas.replace('data:image/png;base64',"");
```
Comment (songz) [inline Meme/express.js]: I prefer this to be done in the client side (browser). Its more efficient this way.

Comment (allopez7) [reply Meme/express.js]: I'm not sure how to do that. This is the way you taught us.

File: Meme/express.js [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+
+  fs.appendFile(textFilePath, chatMessage,()=>{});
+  res.send('append comment to /book.txt');
+}
+app.post('/send', appendTextToFile);
+
+const createImageMeme = (req, res) =>{
+  const imagePath = `/home/alopez/garagescript/Meme/public/${req.body.name}.png`;
+  const imageData = req.body.data;
+  const imageMeme = `/home/alopez/garagescript/Meme/public/${req.body.name}.png`;
```
Comment (songz) [inline Meme/express.js]: y did you create 2 variables for the same thing?

File: Meme/express.js [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+app.post('/send', appendTextToFile);
+
+const createImageMeme = (req, res) =>{
+  const imageData = req.body.data;
+  const imageMeme = `/home/alopez/garagescript/Meme/public/${req.body.name}.png`; 
+
+  fs.writeFile(imageMeme, imageData,'base64',() =>{
+    gm(imageMeme).fontSize(40).drawText(50,50,req.body.comment).write(imageMeme,()=>{
+   });
+});
```
Comment (songz) [inline Meme/express.js]: indentation!

File: Meme/public/index.html [hunk 16]
```diff
@@ -0,0 +1,84 @@
...
+    const button = document.getElementById('submit');
+    button.onclick = () =>{
+      const name = document.getElementById('name').value;
+      const comment = document.getElementById('comment').value;
+      const httpReq = new XMLHttpRequest();
+      httpReq.open('POST','/send');
+      httpReq.setRequestHeader('content-type','application/json');
+      httpReq.send(
+        JSON.stringify({
+          name: name, comment: comment
```
Comment (songz) [inline Meme/public/index.html]: `{name, comment}`

File: Meme/public/index.html [hunk 17]
```diff
@@ -0,0 +1,84 @@
...
+  const comment = document.getElementById('comment').value;
+  const context = canvas.getContext('2d');
+  const httpReq = new XMLHttpRequest();
+  context.drawImage(video,0,0);
+  const dataURL = canvas.toDataURL('image/png').replace('data:image/png;base64',"");
+  
+  httpReq.open('POST','/image');
+  httpReq.setRequestHeader('content-type','application/json');
+  httpReq.send(
+      JSON.stringify(
```
Comment (songz) [inline Meme/public/index.html]: `stringify({ data: dataURL, name, comment})`

**Review by songz — APPROVED** [review]

**Review by saumya665 — APPROVED** [review]

File: Meme/public/index.html [hunk 18]
```diff
@@ -0,0 +1,78 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <meta charset="UTF-8">
+  <title></title>
```
Comment (saumya665) [inline Meme/public/index.html]: nothing inside title?

**Pull request merged**